File: Python/Concurrency/fine_grained_locking.py
import threading

class TicketBookingFineGrained:
    def __init__(self):
        self._locks_lock = threading.Lock()
        self._seat_locks = {}
        self._seat_owners = {}

    def _get_lock(self, seat_id: str) -> threading.Lock:
        with self._locks_lock:
            if seat_id not in self._seat_locks:
                self._seat_locks[seat_id] = threading.Lock()
            return self._seat_locks[seat_id]

    def book_seat(self, seat_id: str, visitor_id: str) -> bool:
        with self._get_lock(seat_id):
            if seat_id in self._seat_owners:
                return False
            self._seat_owners[seat_id] = visitor_id
            return True

class TicketBookingFixed:
    def __init__(self):
        self._seat_locks = {}

    def _get_lock(self, seat_id: str) -> threading.Lock:
        if seat_id not in self._seat_locks:
            self._seat_locks[seat_id] = threading.Lock()
        return self._seat_locks[seat_id]

    def swap_seats(self, visitor1: str, seat1: str,
                   visitor2: str, seat2: str) -> bool:
        # Always acquire locks in consistent order to prevent deadlock
        first = seat1 if seat1 < seat2 else seat2
        second = seat2 if seat1 < seat2 else seat1

        with self._get_lock(first):
            with self._get_lock(second):
                # ... perform swap
                return True

File: CPP concepts/cpp/fine_grained_locking.cpp
#include <mutex>
#include <shared_mutex>
#include <unordered_map>
#include <string>
#include <memory>
using namespace std;

class TicketBookingFineGrained {
private:
    mutex locksMutex;
    // unique_ptr<mutex> is used to point to mutexes created on the heap.
    unordered_map<string, unique_ptr<mutex>> seatLocks;
    unordered_map<string, string> seatOwners;

    mutex& getLock(const string& seatId) {
        lock_guard<mutex> lock(locksMutex);
        if (seatLocks.find(seatId) == seatLocks.end()) {
            seatLocks[seatId] = make_unique<mutex>();
        }
        return *seatLocks[seatId];
    }
public:
    bool bookSeat(const string& seatId, const string& visitorId) {
        lock_guard<mutex> seatLock(getLock(seatId));
        if (seatOwners.find(seatId) != seatOwners.end()) {
            return false;
        }
        seatOwners[seatId] = visitorId;
        return true;
    }
    bool swapSeats(const string& visitor1, const string& seat1, const string& visitor2, const string& seat2) {
        // Always acquire locks in consistent order to prevent deadlock
        const string& first = seat1 < seat2 ? seat1 : seat2;
        const string& second = seat1 < seat2 ? seat2 : seat1;
        // scoped_lock acquires multiple locks without deadlock
        scoped_lock lock(getLock(first), getLock(second));
        // ... perform swap
        return true;
    }
};
